src/DataFramesMeta.jl

```diff
@@ -295,20 +304,40 @@ end
 ##
 ##############################################################################
 
-where(d::AbstractDataFrame, arg) = d[arg, :]
-where(d::AbstractDataFrame, f::Function) = d[f(d), :]
-where(g::GroupedDataFrame, f::Function) = g[Bool[f(x) for x in g]]
+function where_helper(x, args...)
+    t = (fun_to_vec(arg; nolhs = true, gensym_names = true) for arg in args)
+    quote
+        $where($x, $(t...))
+    end
+end
+
+and(x, y) = x .& y
 
-and(x, y) = :($x .& $y)
+function df_to_bool(res)
+    if any(t -> !(t isa Union{Vector{<:Union{Missing, Bool}}, BitArray{1}}), eachcol(res))
+        throw(ArgumentError("All arguments in @where must return a " *
+                            "Vector{<:Union{Missing, Bool} or a BitArray{1}"))
+    end
+
+    return reduce(and, eachcol(res)) .=== true
+end
+
+function where(df::AbstractDataFrame, @nospecialize(args...))
+    res = DataFrames.select(df, args...; copycols = false)
+    tokeep = df_to_bool(res)
+    df[tokeep, :]
+end
 
-function where_helper(d, args...)
-    :($where($d, $(with_anonymous(reduce(and, args)))))
+function where(gd::GroupedDataFrame, @nospecialize(args...))
```

Comment:
why is this defined?

Reply:
I thought we decided that we wanted `where` to work on a GroupedDataFrame.

Reply:
yes - I mixed it up with `orderby`